File: iOS/HandGestureRecognizing/HandGestureRecognizing/HandGestureRecognizing/MotionGate.swift
import Foundation
import HandGestureTypes

/// Pure Phase 1 motion-gate state machine.
///
/// No threading, no callbacks, no I/O — drive it by feeding `HandShot`s in order,
/// inspect the returned `Event` and the public `state` / `bufferCount`. Designed for
/// unit tests: tests construct synthetic `HandShot` sequences with controlled
/// timestamps and assert state transitions.
///
/// `HandGestureRecognizing` owns one `MotionGate` instance and forwards each
/// per-frame handshot to it on its private serial queue.
public final class MotionGate {

    public enum Event: Equatable {
        /// Closed → closed, no transition this frame.
        case stillClosed
        /// Closed → open this frame.
        case opened
        /// Open → open, frame appended to buffer.
        case stillOpen(bufferCount: Int)
        /// Open → closed this frame. The captured buffer is the cycle's payload.
        case cycleEnded(buffer: [HandShot])
    }

    public private(set) var state: MotionGateState = .closed
    public var bufferCount: Int { gateBuffer.count }

    private let config: MotionGateConfig
    private let bufferCap: Int

    private var prevNormalizedCoords: [Float]? = nil
    private var aboveThresholdSince: TimeInterval? = nil
    private var belowThresholdSince: TimeInterval? = nil
    private var gateBuffer: [HandShot] = []

    public init(config: MotionGateConfig, bufferCap: Int = 30) {
        self.config = config
        self.bufferCap = bufferCap
    }

    /// Reset the gate to its initial state. Drops any in-flight buffer.
    public func reset() {
        state = .closed
        prevNormalizedCoords = nil
        aboveThresholdSince = nil
        belowThresholdSince = nil
        gateBuffer.removeAll()
    }

    /// Feed one frame in and observe the state transition this frame produced.
    @discardableResult
    public func process(_ handshot: HandShot) -> Event {
        let now = handshot.timestamp

        // Absent frame → immediate close (with cycle end if was open)
        if handshot.isAbsent {
            let wasOpen = (state == .open)
            let captured = gateBuffer
            gateBuffer.removeAll()
            state = .closed
            prevNormalizedCoords = nil
            aboveThresholdSince = nil
            belowThresholdSince = nil
            return wasOpen ? .cycleEnded(buffer: captured) : .stillClosed
        }

        // Compute energy against previous frame (if any)
        let currCoords = MotionGate.normalize(handshot)
        let energy: Float
        if let prev = prevNormalizedCoords, let curr = currCoords {
            energy = MotionGate.energy(current: curr, previous: prev)
        } else {
            energy = 0
        }
        prevNormalizedCoords = currCoords

        switch state {
        case .closed:
            if energy > config.tOpen {
                if aboveThresholdSince == nil { aboveThresholdSince = now }
                let durationMs = (now - aboveThresholdSince!) * 1000
                if durationMs >= config.kOpenMs {
                    state = .open
                    aboveThresholdSince = nil
                    belowThresholdSince = nil
                    gateBuffer.removeAll()
                    return .opened
                }
            } else {
                aboveThresholdSince = nil
            }
            return .stillClosed

        case .open:
            if gateBuffer.count < bufferCap {
                gateBuffer.append(handshot)
            }

            var shouldClose = false
            if energy < config.tClose {
                if belowThresholdSince == nil { belowThresholdSince = now }
                let durationMs = (now - belowThresholdSince!) * 1000
                if durationMs >= config.kCloseMs {
                    shouldClose = true
                }
            } else {
                belowThresholdSince = nil
            }
            if gateBuffer.count >= bufferCap {
                shouldClose = true
            }

            if shouldClose {
                let captured = gateBuffer
                gateBuffer.removeAll()
                state = .closed
                aboveThresholdSince = nil
                belowThresholdSince = nil
                return .cycleEnded(buffer: captured)
            }
            return .stillOpen(bufferCount: gateBuffer.count)
        }
    }

    // MARK: - Pure helpers (static so tests can call without instantiating a gate)

    /// Wrist-relative, scale-normalised 63-dim coord vector for the shot, or nil if absent/degenerate.
    /// Scale reference is the wrist→middle-finger-MCP distance.
    public static func normalize(_ shot: HandShot) -> [Float]? {
        guard shot.landmarks.count == 21 else { return nil }
        let wrist = shot.landmarks[0]
        let middleMCP = shot.landmarks[9]
        let dx = middleMCP.x - wrist.x
        let dy = middleMCP.y - wrist.y
        let dz = middleMCP.z - wrist.z
        let scaleRef = sqrtf(dx*dx + dy*dy + dz*dz)
        guard scaleRef > 1e-6 else { return nil }
        var coords = [Float](repeating: 0, count: 63)
        for i in 0..<21 {
            let lm = shot.landmarks[i]
            coords[i*3]   = (lm.x - wrist.x) / scaleRef
            coords[i*3+1] = (lm.y - wrist.y) / scaleRef
            coords[i*3+2] = (lm.z - wrist.z) / scaleRef
        }
        return coords
    }

    /// Sum of per-landmark L2 distances between two normalised 63-dim coord vectors.
    public static func energy(current: [Float], previous: [Float]) -> Float {
        var energy: Float = 0
        for i in 0..<21 {
            let dx = current[i*3]   - previous[i*3]
            let dy = current[i*3+1] - previous[i*3+1]
            let dz = current[i*3+2] - previous[i*3+2]
            energy += sqrtf(dx*dx + dy*dy + dz*dz)
        }
        return energy
    }
}

File: core/src/hand_gesture_recognizing/MotionGate.cpp
#include "MotionGate.hpp"
#include <cmath>
#include <algorithm>

MotionGate::MotionGate(const MotionGateConfig& cfg, int buffer_cap)
    : config_(cfg), buffer_cap_(buffer_cap) {}

void MotionGate::reset() {
    state_ = MotionGateState::closed;
    prev_coords_.clear();
    above_threshold_since_.reset();
    below_threshold_since_.reset();
    gate_buffer_.clear();
}

MotionGate::Event MotionGate::process(const cg_handshot& shot) {
    const double now = shot.timestamp;

    // Absent frame → immediate close (cycle end if was open).
    if (shot.is_absent) {
        bool was_open = (state_ == MotionGateState::open);
        std::vector<cg_handshot> captured;
        captured.swap(gate_buffer_);
        state_ = MotionGateState::closed;
        prev_coords_.clear();
        above_threshold_since_.reset();
        below_threshold_since_.reset();
        if (was_open) {
            return {Event::Kind::cycle_ended, 0, std::move(captured)};
        }
        return {Event::Kind::still_closed};
    }

    // Compute energy vs previous frame.
    auto curr_coords = normalize(shot);
    float e = 0.0f;
    if (!prev_coords_.empty() && !curr_coords.empty()) {
        e = energy(curr_coords, prev_coords_);
    }
    prev_coords_ = curr_coords;

    if (state_ == MotionGateState::closed) {
        if (e > config_.t_open) {
            if (!above_threshold_since_) above_threshold_since_ = now;
            double duration_ms = (now - *above_threshold_since_) * 1000.0;
            if (duration_ms >= config_.k_open_ms) {
                state_ = MotionGateState::open;
                above_threshold_since_.reset();
                below_threshold_since_.reset();
                gate_buffer_.clear();
                return {Event::Kind::opened};
            }
        } else {
            above_threshold_since_.reset();
        }
        return {Event::Kind::still_closed};
    }

    // state == open
    if (static_cast<int>(gate_buffer_.size()) < buffer_cap_) {
        gate_buffer_.push_back(shot);
    }

    bool should_close = false;
    if (e < config_.t_close) {
        if (!below_threshold_since_) below_threshold_since_ = now;
        double duration_ms = (now - *below_threshold_since_) * 1000.0;
        if (duration_ms >= config_.k_close_ms) should_close = true;
    } else {
        below_threshold_since_.reset();
    }
    if (static_cast<int>(gate_buffer_.size()) >= buffer_cap_) should_close = true;

    if (should_close) {
        std::vector<cg_handshot> captured;
        captured.swap(gate_buffer_);
        state_ = MotionGateState::closed;
        above_threshold_since_.reset();
        below_threshold_since_.reset();
        return {Event::Kind::cycle_ended, 0, std::move(captured)};
    }
    return {Event::Kind::still_open, static_cast<int>(gate_buffer_.size())};
}

std::vector<float> MotionGate::normalize(const cg_handshot& shot) {
    if (shot.is_absent) return {};
    const auto& lm = shot.landmarks;
    const float wx = lm[0].x, wy = lm[0].y, wz = lm[0].z;
    const float dx = lm[9].x - wx, dy = lm[9].y - wy, dz = lm[9].z - wz;
    const float scale = std::sqrt(dx*dx + dy*dy + dz*dz);
    if (scale < 1e-6f) return {};

    std::vector<float> coords(63);
    for (int i = 0; i < 21; ++i) {
        coords[i*3]   = (lm[i].x - wx) / scale;
        coords[i*3+1] = (lm[i].y - wy) / scale;
        coords[i*3+2] = (lm[i].z - wz) / scale;
    }
    return coords;
}

float MotionGate::energy(const std::vector<float>& cur, const std::vector<float>& prev) {
    float e = 0.0f;
    for (int i = 0; i < 21; ++i) {
        float ddx = cur[i*3]   - prev[i*3];
        float ddy = cur[i*3+1] - prev[i*3+1];
        float ddz = cur[i*3+2] - prev[i*3+2];
        e += std::sqrt(ddx*ddx + ddy*ddy + ddz*ddz);
    }
    return e;
}
